### PR TITLE
Series upgrade

### PR DESCRIPTION
**Pull request by thedac**
Series Upgrade

Utilities for performing series upgrades.
Place holder for future stand alone series upgrade zaza test.
Small change to overlay template to allow stand alone zaza tests.

### CODE REVIEW

File: zaza/utilities/generic.py [hunk 1]
```diff
@@ -165,3 +167,219 @@ def get_yaml_config(config_file):
...
+    """
+    application = unit_name.split('/')[0]
+    juju_utils.prepare_series_upgrade(machine_num, to_series=to_series)
+    logging.info("Watiing for model idleness")
+    model.block_until_all_units_idle()
+    wrap_do_release_upgrade(unit_name, from_series=from_series,
+                            to_series=to_series, files=files,
+                            workaround_script=workaround_script)
+    reboot(unit_name)
+    # Without the sleep model.block_on_all_units_idle returns to early
```
Comment (fnordahl) [inline zaza/utilities/generic.py]: I hit the same problem when changing config and added a helper that looks for `executing` status here:
https://github.com/openstack-charmers/zaza/blob/9eda316756f0e4c5f8ddce979f1761f050292cca/zaza/model.py#L653

I guess we could do something similar for this issue at some point

File: zaza/charm_lifecycle/deploy.py [hunk 2]
```diff
@@ -33,11 +33,7 @@
     'OS_',
     'VIP_RANGE',
 ]
-LOCAL_OVERLAY_TEMPLATE = """
```
[the remark this answers is not in the document]
Comment (thedac) [reply zaza/charm_lifecycle/deploy.py]: It feels weird that an overlay file is necessary. But I do not want to block on this now. We can address that later. I'll revert that bit.

File: zaza/utilities/generic.py (hunk 1 above)
Comment (thedac) [reply zaza/utilities/generic.py]: Used yours as a model. To update async_block_until_unit_wl_status to actually look at individual unit workload state. Update in the next push.

File: zaza/utilities/generic.py [hunk 3]
```diff
@@ -165,3 +166,251 @@ def get_yaml_config(config_file):
...
+    :returns: None
+    :rtype: None
+    """
+    logging.info("Series upgrade {}".format(unit_name))
+    application = unit_name.split('/')[0]
+    logging.info("Prepare series upgrade on {}".format(machine_num))
+    juju_utils.prepare_series_upgrade(machine_num, to_series=to_series)
+    logging.info("Watiing for workload status 'unknown' on {}"
+                 .format(unit_name))
+    model.block_until_unit_wl_status(unit_name, "unknown")
```
Comment (gnuoy) [inline zaza/utilities/generic.py]: Do we still have charms that have not implemented wl status so would have an 'unknown' status before the series upgrade began? Mongo etc

File: zaza/utilities/generic.py [hunk 4]
```diff
@@ -165,3 +166,251 @@ def get_yaml_config(config_file):
...
+def do_release_upgrade(unit_name):
+    """Run do-release-upgrade noninteractive.
+
+    :param unit_name: Unit Name
+    :type unit_name: str
+    :returns: None
+    :rtype: None
+    """
+    logging.info('Upgrading ' + unit_name)
+    # NOTE: It is necessary to run this via juju ssh rather than juju run do to
```
Comment (gnuoy) [inline zaza/utilities/generic.py]: s/do/due/

File: zaza/utilities/juju.py [hunk 5]
```diff
@@ -258,3 +259,77 @@ def leader_get(application, key=''):
...
+    :param machine_num: Machine number
+    :type machine_num: str
+    :param to_series: The series to which to upgrade
+    :type to_series: str
+    :returns: None
+    :rtype: None
+    """
+    juju_model = model.get_juju_model()
+    cmd = ["juju", "upgrade-series", "-m", juju_model,
+           "prepare", machine_num, to_series, "--agree"]
```
Comment (gnuoy) [inline zaza/utilities/juju.py]: I think we should create an upgrade_series method  in zaza/model.py. We can then track the move a subprocess call to a libjuju call.

File: zaza/utilities/juju.py [hunk 6]
```diff
@@ -258,3 +259,77 @@ def leader_get(application, key=''):
...
+    export JUJU_DEV_FEATURE_FLAGS=upgrade-series
+
+    :param machine_num: Machine number
+    :type machine_num: str
+    :returns: None
+    :rtype: None
+    """
+    juju_model = model.get_juju_model()
+    cmd = ["juju", "upgrade-series", "-m", juju_model,
+           "complete", machine_num]
```
Comment (gnuoy) [inline zaza/utilities/juju.py]: Ditto, I think we should create an upgrade_series method in zaza/model.py.

File: zaza/charm_tests/series_upgrade/tests.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+                from_series=from_series,
+                to_series=to_series,
+                origin=origin,
+                workaround_script=workaround_script,
+                files=files)
+
+    def validate_series_upgraded_cloud(self):
+        """Validate post series upgrade."""
+        logging.info("Validate post-series-upgrade: Spin up LTS instance")
+        self.lts.test_launch_small_cirros_instance()
```
[the remark this answers is not in the document]
Comment (thedac) [reply zaza/charm_tests/series_upgrade/tests.py]: Agreed. That function is named incorrectly. Note it is an instance of LTSGuestCreateTest not CirrosGuestCreateTest.

I'll fix in zaza.charm_test.nova.

File: zaza/model.py [hunk 8]
```diff
@@ -1090,25 +1090,27 @@ def f(x):
...
-            lambda: unit.workload_status == status,
-            timeout=timeout
-        )
+    async def _unit_status():
+        app = unit_name.split("/")[0]
+        model_status = await async_get_status()
+        return (model_status.applications[app]['units'][unit_name]
+                ['workload-status']['status'] == status)
+
+    async with run_in_model(model_name):
```
[the remark this answers is not in the document]
Comment (thedac) [reply zaza/model.py]: I'll add more comments. They were buried in the commit message. 

unit.workload_status was actually reporting the application workload status. Using the full status output from model.get_status() gives us unit by unit workload status.

File: zaza/utilities/generic.py (hunk 3 above)
Comment (thedac) [reply zaza/utilities/generic.py]: This is a function of series upgrade. After juju upgrade-series prepare $M $S the juju agents stop and report workload status unknonw and agent lost. Here is an example:

Model              Controller  Cloud/Region             Version    SLA          Timestamp
zaza-d4a4dcf30090  upgrades    serverstack/serverstack  2.5-beta1  unsupported  07:42:13-07:00

App              Version          Status   Scale  Charm            Store  Rev  OS      Charm version  Notes
pc-hacluster                      blocked    2/3  hacluster        local    0  ubuntu  f6e9ee9        
percona-cluster  5.5.37-25.10...  blocked    2/3  percona-cluster  local   45  ubuntu  f6e9ee9        

Unit                Workload     Agent  Machine  Public address  Ports     Message
percona-cluster/0   unknown      lost   0        10.5.0.22       3306/tcp  agent lost, see 'juju show-status-log percona-cluster/0'
  pc-hacluster/2    unknown      lost            10.5.0.22                 agent lost, see 'juju show-status-log pc-hacluster/2'
percona-cluster/1*  maintenance  idle   1        10.5.0.12       3306/tcp  Paused. Use 'resume' action to resume normal service.
  pc-hacluster/0*   maintenance  idle            10.5.0.12                 Paused. Use 'resume' action to resume normal service.
percona-cluster/2   maintenance  idle   2        10.5.0.17       3306/tcp  Paused. Use 'resume' action to resume normal service.
  pc-hacluster/1    maintenance  idle            10.5.0.17                 Paused. Use 'resume' action to resume normal service.

File: zaza/utilities/generic.py [hunk 9]
```diff
@@ -165,3 +166,251 @@ def get_yaml_config(config_file):
     # the pwd.
     logging.info('Using config %s' % (config_file))
     return yaml.load(open(config_file, 'r').read())
+
+
+def series_upgrade_application(application, pause_non_leader_primary=True,
+                               pause_non_leader_subordinate=True,
+                               from_series="trusty", to_series="xenial",
```
[the remark this answers is not in the document]
Comment (thedac) [reply zaza/utilities/generic.py]: Would you prefer no defaults for those?

File: zaza/utilities/generic.py [hunk 10]
```diff
@@ -165,3 +166,251 @@ def get_yaml_config(config_file):
...
+                               files=None, workaround_script=None):
+    """Series upgrade application.
+
+    Wrap all the functionality to handle series upgrade for a given
+    application. Including pausing non-leader units.
+
+    :param application: Name of application to upgrade series
+    :type application: str
+    :param pause_non_leader_primary: Should the non-leader applications be
+                                     paused?
```
[the remark this answers is not in the document]
Comment (thedac) [reply zaza/utilities/generic.py]: Done

File: zaza/utilities/generic.py [hunk 11]
```diff
@@ -165,3 +166,251 @@ def get_yaml_config(config_file):
...
+    :type pause_non_leader_primary: bool
+    :param pause_non_leader_subordinate: Should the non-leader subordinate
+                                         hacluster applications be paused?
```
[the remark this answers is not in the document]
Comment (thedac) [reply zaza/utilities/generic.py]: Done

File: zaza/utilities/generic.py (hunk 4 above)
Comment (thedac) [reply zaza/utilities/generic.py]: Done

File: zaza/utilities/juju.py (hunk 5 above)
Comment (thedac) [reply zaza/utilities/juju.py]: Working on it

File: zaza/utilities/juju.py (hunk 6 above)
Comment (thedac) [reply zaza/utilities/juju.py]: Working on it

File: zaza/utilities/generic.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (thedac) [reply zaza/utilities/generic.py]: Done

File: zaza/model.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (thedac) [reply zaza/model.py]: Done

**Pull request merged**